include/istio/utils/attributes_builder.h

```diff
@@ -67,38 +67,38 @@ class AttributesBuilder {
     time_stamp->set_nanos(nanos % 1000000000);
   }
 
-  void AddDuration(const std::string& key,
-                   const std::chrono::nanoseconds& value) {
+  void AddDuration(const std::string &key,
+                   const std::chrono::nanoseconds &value) {
     auto duration =
         (*attributes_->mutable_attributes())[key].mutable_duration_value();
     duration->set_seconds(value.count() / 1000000000);
     duration->set_nanos(value.count() % 1000000000);
   }
 
-  void AddStringMap(const std::string& key,
-                    const std::map<std::string, std::string>& string_map) {
+  void AddStringMap(const std::string &key,
+                    const std::map<std::string, std::string> &string_map) {
     if (string_map.size() == 0) {
       return;
     }
     auto entries = (*attributes_->mutable_attributes())[key]
                        .mutable_string_map_value()
                        ->mutable_entries();
     entries->clear();
-    for (const auto& map_it : string_map) {
+    for (const auto &map_it : string_map) {
       (*entries)[map_it.first] = map_it.second;
     }
   }
 
-  void AddProtoStructStringMap(const std::string& key,
-                               const google::protobuf::Struct& struct_map) {
+  void AddProtoStructStringMap(const std::string &key,
+                               const google::protobuf::Struct &struct_map) {
     if (struct_map.fields().empty()) {
       return;
     }
     auto entries = (*attributes_->mutable_attributes())[key]
                        .mutable_string_map_value()
                        ->mutable_entries();
     entries->clear();
-    for (const auto& field : struct_map.fields()) {
+    for (const auto &field : struct_map.fields()) {
```

Comment:
Please add:  
if entries is still empty after the loop,  remove the key.  

I like to have a option for filters not to dump data to mixer.  Or another way,  if filters want to dump data to mixer, they have to say so.   We should not blindly dump everything to mixer.  I have concern about performance, and security

Reply:
This is only in the report path. Dynamic metadata is strictly for logging.
See https://github.com/envoyproxy/envoy/commit/700b7d09f400ada6fa5b2202d2588441b6f7cb66
In other words, even in Envoy, all dynamic metadata is already dumped on the wire, as is, just like static metadata. If filters store sensitive information, then there are bigger concerns including access logs, header values, etc. which are also being sent to mixer. So, lets have a separate discussion for that.

Reply:
I agree with Shriram on the security front, I'm not concerned.

I do wonder about the end-to-end perf impact. Can we measure the cost on proxy and Mixer CPU usage?

Reply:
No matter what we do, there is no way to tell other external filters that they need to log things for Istio mixer. Second, dynamic metadata is not 1000s of kbs of data. Its very very small bits of info like name of resource, or some key claim type, etc. This is the *only* way the mixer has, if it wishes to be able to log info about other OSS codecs like mongo/redis/kafka in Envoy. Otherwise, we will still be stuck with opaque TCP attributes like bytes sent, received, etc., while people prefer to turn off mixer and use envoy gRPC access logging to get richer metrics. :)

Reply:
>logging is for debugging 
???
Access log is most certainly not debugging. In fact, this PR came from Google: https://github.com/envoyproxy/envoy/commit/700b7d09f400ada6fa5b2202d2588441b6f7cb66 and this is adding dynamic metadata to access logs streamed over gRPC. 
Remember, I can always choose to overload the mixer by sending very very large HTTP headers, auth keys, etc. or several 100s of headers. So, the argument about performance is moot. We can potentially agree on some common prefix but that requires discussion with envoy folks as well, things like mongo/redis that generate dynamic metadata need to be modified to use this agreed upon prefix. This will take time. Do I have to be wait until then?